qiskit/providers/ibmq/jupyter/dashboard/watcher_monitor.py

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A module of widgets for job monitoring"""
+import sys
+import time
+import threading
+from ...utils.converters import utc_to_local
+
+
+def _job_monitor(job, status, watcher):
+    """Monitor the status of a IBMQJob instance.
+
+    Args:
+        job (BaseJob): Job to monitor.
+        status (Enum): Job status.
+        watcher (JobWatcher): Job watcher instance
+    """
+    thread = threading.Thread(target=_job_checker, args=(job, status, watcher))
+    thread.start()
+
+
+def _job_checker(job, status, watcher):
+    """A simple job status checker
+
+    Args:
+        job (BaseJob): The job to check.
+        status (Enum): Job status.
+        watcher (JobWatcher): Job watcher instance
+
+    """
+    prev_status_name = None
+    prev_queue_pos = None
+    interval = 2
+    exception_count = 0
+
+    while status.name not in ['DONE', 'CANCELLED', 'ERROR']:
+        time.sleep(interval)
+        try:
+            status = job.status()
```

Comment:
Instead of doing `job.status()` in a loop, you can now use `wait_for_final_state()`, which uses websockets and thus generates far less traffic to the API. This can be in a separate PR though.

Reply:
I am not waiting for the final state here, but rather am looking for changes in state so as to update the job monitor accordingly.

Reply:
The `while` on line 48 says it is checking for job status until a final state is reached. You can supply a callback function to `wait_for_final_state()` that updates the job monitor as the status changes.

Reply:
Wait for the final state does not do an adaptive time interval like the current checker does. Also, why does job.status not use websockets?

Reply:
True, I should add that feature (not for this PR but later). 
`job.status()` doesn't use websockets because it returns immediately, which defeats the purpose of using a long running websocket.

Reply:
`job.status` and `job.wait_for_final_state` both call `job._get_status_position` with the same arguments to get the running status.  How is it that the latter uses websockets, while the former does not?

Reply:
`job._get_status_position()` is used to interpret the API response, not to communicate with the API. `job.status()` calls `._api.job_status()`, which does a 1 time call to the API that returns immediately. `job.wait_for_final_state()` calls `._api.job_final_status()`, which keeps the connection open (via websockets) until the job is done.

Reply:
So another issue with the callback is that it does not keep state from one call to the next.  So things like changing HTML status fields only if the status has changed cannot be supported because there is no way to keep track of the status from one callback to the next.

Reply:
Yeah as mentioned before, invoking the callback only if something changed (status or queue position etc) would be a nice feature to add to `wait_for_final_state`. But that'd be a separate PR. The job watcher will just have to continue polling (unless you can extract the previous state from the widget somehow, but updating `wait_for_final_state` seems to be the better solution anyway).

Reply:
Yes, but as it stands now I could be doing a ton of widget refreshing for no reason because the callback does not support it.  I prefer to wait for the callback to be more generalized before moving all these routines which have been tested for many months now, over to it.  But I agree it should be done.